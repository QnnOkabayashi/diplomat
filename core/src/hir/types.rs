//! Types that can be exposed in Diplomat APIs.

use super::{paths, PrimitiveType, TypeContext, TypeLifetime};

/// Type that may be used as an output.
pub enum ReturnableType {
    Type(Type),
    OutType(OutType),
}

/// Type that can only be used as an output.
pub enum OutType {
    Primitive(PrimitiveType),
    Opaque(paths::OutOpaque),
    Struct(paths::ReturnableStruct),
    Enum(paths::Enum),
    Slice(Slice),
}

/// Type that may be used as input or output.
pub enum Type {
    Primitive(PrimitiveType),
    Opaque(paths::Opaque),
    Struct(paths::Struct),
    Enum(paths::Enum),
    Slice(Slice),
}

/// Type that can appear in the `self` position.
pub enum SelfType {
    Opaque(paths::SelfOpaque),
    Struct(paths::Struct),
    Enum(paths::Enum),
}

#[derive(Copy, Clone)]
pub enum Slice {
    /// A string slice, e.g. `&str`.
    Str(TypeLifetime),

    /// A primitive slice, e.g. `&mut [u8]`.
    Primitive(Ref, PrimitiveType),
}

#[derive(Copy, Clone)]
pub enum Mutability {
    Mutable,
    Immutable,
}

#[derive(Copy, Clone)]
pub struct Ref {
    pub lifetime: TypeLifetime,
    pub mutability: Mutability,
}

impl Type {
    /// Return the number of fields and leaves that will show up in the [`LifetimeTree`]
    /// returned by [`Param::lifetime_tree`] and [`ParamSelf::lifetime_tree`].
    ///
    /// This method is used to calculate how much space to allocate upfront.
    pub(super) fn field_leaf_lifetime_counts(&self, tcx: &TypeContext) -> (usize, usize) {
        match self {
            Type::Struct(ty) => ty.resolve(tcx).fields.iter().fold((1, 0), |acc, field| {
                let inner = field.ty.field_leaf_lifetime_counts(tcx);
                (acc.0 + inner.0, acc.1 + inner.1)
            }),
            Type::Opaque(_) | Type::Slice(_) => (0, 1),
            Type::Primitive(_) | Type::Enum(_) => (0, 0),
        }
    }
}

impl Slice {
    /// Returns the [`TypeLifetime`] contained in either the `Str` or `Primitive`
    /// variant.
    pub fn lifetime(&self) -> &TypeLifetime {
        match self {
            Slice::Str(lifetime) => lifetime,
            Slice::Primitive(reference, _) => &reference.lifetime,
        }
    }
}
