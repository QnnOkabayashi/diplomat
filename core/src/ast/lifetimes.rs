use proc_macro2::Span;
use quote::{quote, ToTokens};
use serde::{Deserialize, Serialize};
use std::fmt;

use super::Ident;

/// A named lifetime, e.g. `'a`.
///
/// # Invariants
///
/// Cannot be `'static` or `'_`, use [`Lifetime`] to represent those instead.
#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, PartialOrd, Ord)]
#[serde(transparent)]
pub struct NamedLifetime(Ident);

impl<'de> Deserialize<'de> for NamedLifetime {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        // Special `Deserialize` impl to ensure invariants.
        let named = Ident::deserialize(deserializer)?;
        if named.as_str() == "static" {
            panic!("cannot be static");
        }
        Ok(NamedLifetime(named))
    }
}

impl From<&syn::Lifetime> for NamedLifetime {
    fn from(lt: &syn::Lifetime) -> Self {
        Lifetime::from(lt).to_named().expect("cannot be static")
    }
}

impl From<&NamedLifetime> for NamedLifetime {
    fn from(this: &NamedLifetime) -> Self {
        this.clone()
    }
}

impl PartialEq<syn::Lifetime> for NamedLifetime {
    fn eq(&self, other: &syn::Lifetime) -> bool {
        other.ident == self.0.as_str()
    }
}

impl fmt::Display for NamedLifetime {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "'{}", self.0)
    }
}

impl ToTokens for NamedLifetime {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        use proc_macro2::{Punct, Spacing};
        Punct::new('\'', Spacing::Joint).to_tokens(tokens);
        self.0.to_tokens(tokens);
    }
}

/// A lifetime dependency graph used for tracking which lifetimes outlive,
/// and are outlived by, other lifetimes.
///
/// It is similar to [`syn::LifetimeDef`], except it can also track lifetime
/// bounds defined in the `where` clause.
#[derive(Clone, PartialEq, Eq, Hash, Debug, Default)]
pub struct LifetimeEnv {
    edges: Vec<LifetimeEdges>,
}

impl LifetimeEnv {
    /// Collect all lifetimes that live _at least_ as long as _any_ of the
    /// provided `root_lifetimes`.
    ///
    /// This method performs a depth-first search through the lifetime graph,
    /// following edges from short lifetimes to longer lifetimes. Since lifetime
    /// bounds are transitive, this traversal ensures that every visited lifetime
    /// will outlive the root lifetime, and furthermore, DFS ensures that _all_
    /// provably longer lifetimes than the root lifetime are visited. This method
    /// also tracks which lifetimes have already been visited, making it robust
    /// to cycles.
    pub fn outlives<'i, 's, I>(&'s self, root_lifetimes: I) -> Vec<&'s NamedLifetime>
    where
        I: IntoIterator<Item = &'i NamedLifetime>,
    {
        let iter = root_lifetimes.into_iter();
        let mut outlives = Vec::with_capacity(iter.size_hint().1.unwrap_or(0));

        // Track visited lifetimes to avoid cycles.
        let mut visited = vec![false; self.edges.len()];

        iter.filter_map(|named| {
            // Lifetimes that don't have a position aren't in the graph,
            // and thus have no known lifetimes that outlive them.
            self.edges.iter().position(|edge| edge.lifetime == *named)
        })
        .for_each(|root_id| {
            dfs(root_id, &self.edges[..], &mut outlives, &mut visited[..]);
        });

        /// Perform recursive DFS on a lifetime's sub-lifetimes.
        fn dfs<'a>(
            id: usize,
            edges: &'a [LifetimeEdges],
            outlives: &mut Vec<&'a NamedLifetime>,
            visited: &mut [bool],
        ) {
            // Note: all of these indexings SHOULD be valid because
            // `visited.len() == edges.len()`, and the ids come from
            // calling `Iterator::position` on `edges`, which never shrinks.
            // So we should be able to change these to `get_unchecked`...
            if !visited[id] {
                visited[id] = true;

                let edge = &edges[id];
                outlives.push(&edge.lifetime);
                for &longer_id in edge.longer.iter() {
                    dfs(longer_id, edges, outlives, visited);
                }
            }
        }

        outlives
    }

    /// Add the lifetimes from generic parameters and where bounds.
    pub fn extend_generics(&mut self, generics: &syn::Generics) {
        // It's the responsibility of the validity checks to ensure
        // that all the generics are lifetime bounds, not generic types.
        self.extend_from_parts(generics.lifetimes().map(|def| (&def.lifetime, &def.bounds)));
        if let Some(ref where_clause) = generics.where_clause {
            self.extend_from_parts(where_clause.predicates.iter().map(|pred| match pred {
                syn::WherePredicate::Type(_) => panic!("trait bounds are unsupported"),
                syn::WherePredicate::Lifetime(pred) => (&pred.lifetime, &pred.bounds),
                syn::WherePredicate::Eq(_) => panic!("eq bounds are unsupported"),
            }));
        }
    }

    /// Returns the number of lifetimes in the graph.
    pub fn len(&self) -> usize {
        self.edges.len()
    }

    /// Returns `true` if the graph contains no lifetimes.
    pub fn is_empty(&self) -> bool {
        self.edges.is_empty()
    }

    /// `<'a, 'b, 'c>`
    ///
    /// Write the existing lifetimes, excluding bounds, as generic parameters.
    ///
    /// To include lifetime bounds, use [`LifetimeEnv::lifetime_defs_to_tokens`].
    pub fn lifetimes_to_tokens(&self) -> proc_macro2::TokenStream {
        if self.is_empty() {
            return quote! {};
        }

        let lifetimes = self.edges.iter().map(|edge| &edge.lifetime);
        quote! { <#(#lifetimes),*> }
    }

    /// Helper function for `LifetimeGraph::extend_from_parts`,
    /// which allows for getting the id of an `Edge` associated with a lifetime.
    ///
    /// Returns `Ok` if the lifetime was already in the graph, and `Err` if a
    /// new lifetime had to be inserted.
    fn id_or_insert<L>(&mut self, lifetime: &L) -> Result<usize, usize>
    where
        NamedLifetime: PartialEq<L> + for<'a> From<&'a L>,
    {
        // We have to index because we need to call this twice in the same scope
        // in `extend_from_parts`, otherwise I would've just returned `&mut Edge`.
        if let Some(idx) = self
            .edges
            .iter()
            .position(|edge| &edge.lifetime == lifetime)
        {
            // The edge for this lifetime already exists.
            Ok(idx)
        } else {
            // The edge doesn't exist yet, create it and return its id.
            let id = self.edges.len();
            self.edges.push(LifetimeEdges {
                lifetime: lifetime.into(),
                shorter: vec![],
                longer: vec![],
            });
            Err(id)
        }
    }

    /// Extends the `LifetimeEnv` from an iterator of parts found in lifetime bounds.
    ///
    /// # Panics
    ///
    /// This method panics if, by the time all the parts are read in, there are
    /// any lifetimes that have been used as bounds but _not_ declared. This should
    /// only be of concern during deserialization.
    // These trait bounds look intimidating, here's what they do:
    //  * L: A `NamedLifetime`-like type
    //  * B: An iterator of L's
    //  * I: An iterator of (L, B), basically just each lifetime and its bounds.
    // The reason we have this is so we can extend from `syn` parts, or from
    // `&NamedLifetime`s for when we're deserializing.
    pub fn extend_from_parts<'a, I, L, B>(&mut self, iter: I)
    where
        NamedLifetime: PartialEq<L> + for<'b> From<&'b L>,
        L: 'a,
        B: 'a + IntoIterator<Item = &'a L>,
        I: IntoIterator<Item = (&'a L, B)>,
    {
        // Lifetimes that have been used as bounds, but not actually declared
        let mut undeclared = vec![];

        for (lifetime, bounds) in iter {
            // Since all the indices come from me, we could theoretically
            // change these to `get_unchecked`. But unsafe is spooky :/
            let long_id = match self.id_or_insert(lifetime) {
                Ok(id) => {
                    if let Some(undeclared_id) = undeclared
                        .iter()
                        .position(|undeclared_id| *undeclared_id == id)
                    {
                        // This lifetime was previously used as a bound, but now
                        // we're declaring it.
                        undeclared.swap_remove(undeclared_id);
                    }
                    id
                }
                Err(id) => id,
            };

            for bound in bounds {
                let short_id = self.id_or_insert(bound).unwrap_or_else(|id| {
                    // A lifetime has been used as a bound but hasn't been declared yet.
                    undeclared.push(id);
                    id
                });
                // This doesn't catch repeats. But that doesn't break anything
                // and it won't slow down the DFS, so...
                self.edges[short_id].longer.push(long_id);
                self.edges[long_id].shorter.push(short_id);
            }
        }

        if let Some((&first, rest)) = undeclared.split_first() {
            // There's at least one lifetime that's undeclared.
            use std::fmt::Write;
            let mut msg = format!(
                "used undeclared lifetimes in bounds: [{}",
                self.edges[first].lifetime
            );
            for &id in rest {
                write!(msg, ", {}", self.edges[id].lifetime).unwrap();
            }
            msg.write_char(']').unwrap();
            panic!("{}", msg)
        }
    }
}

impl fmt::Display for LifetimeEnv {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.to_token_stream().fmt(f)
    }
}

impl From<&syn::Generics> for LifetimeEnv {
    fn from(generics: &syn::Generics) -> Self {
        let mut this = Self::default();
        this.extend_generics(generics);
        this
    }
}

impl ToTokens for LifetimeEnv {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        for edge in self.edges.iter() {
            let lifetime = &edge.lifetime;
            if edge.shorter.is_empty() {
                tokens.extend(quote! { #lifetime, });
            } else {
                let bounds = edge.shorter.iter().map(|&id| &self.edges[id].lifetime);
                tokens.extend(quote! { #lifetime: #(#bounds)+*, });
            }
        }
    }
}

/// Serialize a [`LifetimeEnv`] as a map from lifetimes to their bounds.
impl Serialize for LifetimeEnv {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        use serde::ser::SerializeMap;
        let mut seq = serializer.serialize_map(Some(self.len()))?;

        for edge in self.edges.iter() {
            /// Helper type for serializing bounds.
            struct Bounds<'a> {
                ids: &'a [usize],
                edges: &'a [LifetimeEdges],
            }

            impl<'a> Serialize for Bounds<'a> {
                fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                where
                    S: serde::Serializer,
                {
                    use serde::ser::SerializeSeq;
                    let mut seq = serializer.serialize_seq(Some(self.ids.len()))?;
                    for &id in self.ids {
                        seq.serialize_element(&self.edges[id].lifetime)?;
                    }
                    seq.end()
                }
            }

            seq.serialize_entry(
                &edge.lifetime,
                &Bounds {
                    ids: &edge.shorter[..],
                    edges: &self.edges,
                },
            )?;
        }
        seq.end()
    }
}

impl<'de> Deserialize<'de> for LifetimeEnv {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        use std::collections::BTreeMap;

        let m: BTreeMap<NamedLifetime, Vec<NamedLifetime>> =
            Deserialize::deserialize(deserializer)?;

        let mut this = LifetimeEnv::default();
        this.extend_from_parts(m.iter());
        Ok(this)
    }
}

/// A lifetime, along with ptrs to all lifetimes that are explicitly
/// shorter/longer than it.
///
/// This type is internal to [`LifetimeGraph`]- the ptrs are stored as `usize`s,
/// meaning that they may be invalid if a `LifetimeEdges` is created in one
/// `LifetimeGraph` and then used in another.
#[derive(Clone, PartialEq, Eq, Hash, Debug)]
struct LifetimeEdges {
    /// The name of the lifetime.
    lifetime: NamedLifetime,

    /// Pointers to all lifetimes that this lives _at least_ as long as.
    ///
    /// Note: This doesn't account for transitivity.
    shorter: Vec<usize>,

    /// Pointers to all lifetimes that live _at least_ as long as this.
    ///
    /// Note: This doesn't account for transitivity.
    longer: Vec<usize>,
}

/// A lifetime, analogous to [`syn::Lifetime`].
#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
pub enum Lifetime {
    /// The `'static` lifetime.
    Static,

    /// A named lifetime, like `'a`.
    Named(NamedLifetime),

    /// An elided lifetime.
    Anonymous,
}

impl Lifetime {
    pub fn to_named(self) -> Option<NamedLifetime> {
        if let Lifetime::Named(named) = self {
            return Some(named);
        }
        None
    }
}

impl fmt::Display for Lifetime {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Lifetime::Static => "'static".fmt(f),
            Lifetime::Named(ref named) => named.fmt(f),
            Lifetime::Anonymous => "'_".fmt(f),
        }
    }
}

impl ToTokens for Lifetime {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            Lifetime::Static => syn::Lifetime::new("'static", Span::call_site()).to_tokens(tokens),
            Lifetime::Named(ref s) => s.to_tokens(tokens),
            Lifetime::Anonymous => syn::Lifetime::new("'_", Span::call_site()).to_tokens(tokens),
        };
    }
}

impl From<&syn::Lifetime> for Lifetime {
    fn from(lt: &syn::Lifetime) -> Self {
        if lt.ident == "static" {
            Self::Static
        } else {
            Self::Named(NamedLifetime((&lt.ident).into()))
        }
    }
}

impl From<&Option<syn::Lifetime>> for Lifetime {
    fn from(lt: &Option<syn::Lifetime>) -> Self {
        lt.as_ref().map(Into::into).unwrap_or(Self::Anonymous)
    }
}

impl Lifetime {
    /// Converts the [`Lifetime`] back into an AST node that can be spliced into a program.
    pub fn to_syn(&self) -> Option<syn::Lifetime> {
        match *self {
            Self::Static => Some(syn::Lifetime::new("'static", Span::call_site())),
            Self::Anonymous => None,
            Self::Named(ref s) => Some(syn::Lifetime::new(&s.to_string(), Span::call_site())),
        }
    }
}
