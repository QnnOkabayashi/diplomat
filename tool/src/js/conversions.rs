use diplomat_core::Env;
use std::fmt;
use std::fmt::Write;

use diplomat_core::ast;

use super::display;
use super::types::{return_type_form, ReturnTypeForm};
use crate::layout;

/// TODO: docs
#[allow(clippy::ptr_arg)] // false positive, rust-clippy#8463, fixed in 1.61
pub fn gen_value_js_to_rust(
    param_name: &ast::Ident,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    borrowed_lifetimes: &[&ast::NamedLifetime],
    pre_logic: &mut Vec<String>,
    invocation_params: &mut Vec<String>,
    post_logic: &mut Vec<String>,
) {
    match typ {
        ast::TypeName::StrReference(lifetime) | ast::TypeName::PrimitiveSlice(lifetime, ..) => {
            let js_param_name = if let ast::TypeName::PrimitiveSlice(.., prim) = typ {
                let js_param_name = ast::Ident::from(format!("{}_diplomat_slice", param_name));
                pre_logic.push(format!(
                    "let {js_param_name} = diplomatRuntime.DiplomatBuf.slice(wasm, {param_name}, {align});",
                    align = layout::primitive_size_alignment(*prim).align()
                ));
                js_param_name
            } else {
                let js_param_name = ast::Ident::from(format!("{}_diplomat_str", param_name));
                pre_logic.push(format!(
                    "let {js_param_name} = diplomatRuntime.DiplomatBuf.str(wasm, {param_name});"
                ));
                js_param_name
            };

            invocation_params.push(format!("{js_param_name}.ptr"));
            invocation_params.push(format!("{js_param_name}.size"));

            match lifetime {
                ast::Lifetime::Named(ref named) if borrowed_lifetimes.contains(&named) => {
                    // TODO: don't create an edge if this is a str/slice,
                    // since they actually copy and thus don't rely on the buffer.
                    post_logic.push(format!(
                        "diplomat_out.__{param_name}_lifetime_guard = {js_param_name};",
                    ));
                }
                _ => {
                    post_logic.push(format!("{js_param_name}.free();"));
                }
            }
        }
        ast::TypeName::Primitive(ast::PrimitiveType::char) => {
            // we use the spread operator here to count codepoints
            // codePointAt() does not return surrogate pairs if there are multiple
            invocation_params.push(format!(
                "diplomatRuntime.extractCodePoint({p}, '{p}')",
                p = param_name
            ));
        }
        ast::TypeName::Box(_) => {
            invocation_params.push(format!("{}.underlying", param_name));
        }
        ast::TypeName::Reference(lifetime, _mut, _typ) => {
            match lifetime {
                ast::Lifetime::Named(named) if borrowed_lifetimes.contains(&named) => {
                    // We want to hold the reference here
                }
                _ => {}
            }
            invocation_params.push(format!("{param_name}.underlying"));
        }
        ast::TypeName::Named(path_type) => {
            match path_type.resolve(in_path, env) {
                ast::CustomType::Struct(struct_type) => {
                    // == THE GAMEPLAN ==
                    // do `Struct::borrowed_lifetimes`, which takes
                    // a filter that tells which indices of lifetimes we need.
                    // Let that filter take `i`, an index. do `path_type.lifetimes[i]`
                    // to get what the lifetime is in the current `LifetimeEnv`
                    // (not the `Struct`'s env!), and then see if that lifetime
                    // in `borrowed_lifetimes`. If it is,
                    // have the filter return `true`, otherwise `false`.
                    // The `Struct::borrowed_lifetimes` method will then return
                    // a list of all the relevant lifetimes in the `Struct`'s
                    // `LifetimeEnv` scope, which we can pass down to further
                    // recursive calls.
                    // Once we get this all working, we can start thinking about indices.
                    // ==================

                    // Shadow the borrowed lifetime from this invocation so they're
                    // ready for the recursive invocation.
                    let borrowed_lifetimes = struct_type.borrowed_lifetimes(|i| {
                        match path_type.lifetimes.get(i) {
                            Some(ast::Lifetime::Static) => false,
                            Some(ast::Lifetime::Named(ref named)) => {
                                borrowed_lifetimes.contains(&named)
                            }
                            Some(ast::Lifetime::Anonymous) | None => {
                                panic!("lifetime elision is unsupported, {} missing lifetime at index {}", path_type, i)
                            }
                        }
                    });

                    for (field_name, field_type, _) in struct_type.fields.iter() {
                        let field_extracted_name =
                            format!("diplomat_{}_extracted_{}", struct_type.name, field_name)
                                .into();
                        pre_logic.push(format!(
                            "const {} = {}[\"{}\"];",
                            field_extracted_name, param_name, field_name
                        ));

                        gen_value_js_to_rust(
                            &field_extracted_name,
                            field_type,
                            in_path,
                            env,
                            &borrowed_lifetimes[..],
                            pre_logic,
                            invocation_params,
                            post_logic,
                        );
                    }
                }
                ast::CustomType::Enum(enm) => {
                    invocation_params.push(format!("{}_js_to_rust[{}]", enm.name, param_name));
                }

                ast::CustomType::Opaque(_) => {
                    panic!("Opaque types cannot be sent as values");
                }
            }
        }
        _ => invocation_params.push(param_name.to_string()),
    }
}

/// TODO: docs
pub fn gen_value_rust_to_js<W: fmt::Write>(
    value_expr: &str,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    borrowed_params: &ast::BorrowedParams,
    env: &Env,
    out: &mut W,
) -> fmt::Result {
    match typ {
        ast::TypeName::Named(path_type) => match path_type.resolve(in_path, env) {
            ast::CustomType::Struct(strct) => {
                let strct_size_align = layout::type_size_alignment(typ, in_path, env);
                let needs_buffer = return_type_form(typ, in_path, env);
                if needs_buffer != ReturnTypeForm::Complex {
                    todo!("Receiving structs that don't need a buffer: {}", strct.name)
                    // recursively call on single field
                }
                write!(
                    out,
                    "{}",
                    display::iife(|mut f| {
                        writeln!(
                            f,
                            "const diplomat_receive_buffer = wasm.diplomat_alloc({}, {});",
                            strct_size_align.size(),
                            strct_size_align.align(),
                        )?;
                        writeln!(f, "{};", value_expr)?;
                        writeln!(
                            f,
                            "const out = new {}(diplomat_receive_buffer);",
                            strct.name
                        )?;

                        writeln!(
                            f,
                            "wasm.diplomat_free(diplomat_receive_buffer, {size}, {align});",
                            size = strct_size_align.size(),
                            align = strct_size_align.align()
                        )?;

                        writeln!(f, "return out;")
                    })
                )?;
            }

            ast::CustomType::Enum(enm) => {
                write!(out, "{}_rust_to_js[{}]", enm.name, value_expr)?;
            }

            ast::CustomType::Opaque(_) => {
                panic!("Opaque types cannot be used in value position")
            }
        },

        ast::TypeName::Box(underlying) => {
            gen_rust_reference_to_js(
                underlying.as_ref(),
                in_path,
                value_expr,
                borrowed_params,
                env,
                out,
            )?;
        }

        ast::TypeName::Option(_) => {
            gen_rust_reference_to_js(typ, in_path, value_expr, borrowed_params, env, out)?;
        }

        ast::TypeName::Result(ok, err) => {
            let (ok_offset, result_size_align) =
                layout::result_ok_offset_size_align(ok, err, in_path, env);
            let needs_buffer = return_type_form(typ, in_path, env) == ReturnTypeForm::Complex;
            write!(
                out,
                "{}",
                display::iife(|mut f| {
                    let (size, align) = (result_size_align.size(), result_size_align.align());
                    if needs_buffer {
                        writeln!(
                            f,
                            "const diplomat_receive_buffer = wasm.diplomat_alloc({size}, {align});"
                        )?;
                        writeln!(f, "{};", value_expr)?;
                        writeln!(
                            f,
                            "const is_ok = {};",
                            display::expr(|mut f| {
                                gen_rust_reference_to_js(
                                    &ast::TypeName::Primitive(ast::PrimitiveType::bool),
                                    in_path,
                                    &format!("diplomat_receive_buffer + {}", ok_offset),
                                    borrowed_params,
                                    env,
                                    &mut f,
                                )
                            })
                        )?;
                        writeln!(
                            f,
                            "if (is_ok) {is_true} else {is_false}",
                            is_true = display::block(|mut f| {
                                writeln!(
                                    f,
                                    "const ok_value = {};",
                                    display::expr(|mut f| {
                                        gen_rust_reference_to_js(
                                            ok.as_ref(),
                                            in_path,
                                            "diplomat_receive_buffer",
                                            borrowed_params,
                                            env,
                                            &mut f,
                                        )
                                    })
                                )?;
                                writeln!(
                                    f,
                                    "wasm.diplomat_free(diplomat_receive_buffer, {size}, {align});"
                                )?;
                                writeln!(f, "return ok_value;")
                            }),
                            is_false = display::block(|mut f| {
                                writeln!(
                                    f,
                                    "const throw_value = {};",
                                    display::expr(|mut f| {
                                        gen_rust_reference_to_js(
                                            err.as_ref(),
                                            in_path,
                                            "diplomat_receive_buffer",
                                            borrowed_params,
                                            env,
                                            &mut f,
                                        )
                                    })
                                )?;
                                writeln!(
                                    f,
                                    "wasm.diplomat_free(diplomat_receive_buffer, {size}, {align});"
                                )?;
                                writeln!(f, "throw new diplomatRuntime.FFIError(throw_value);")
                            })
                        )
                    } else {
                        writeln!(f, "const is_ok = {} == 1;", value_expr)?;
                        writeln!(
                            f,
                            "if (!is_ok) {}",
                            display::block(|mut f| {
                                writeln!(f, "throw new diplomatRuntime.FFIError({{}});")
                            })
                        )
                    }
                })
            )?;
        }

        ast::TypeName::Primitive(_prim) => {
            // TODO(#63): wrap with appropriate types for large widths
            write!(out, "{}", value_expr)?;
        }

        ast::TypeName::Reference(.., underlying) => {
            gen_rust_reference_to_js(
                underlying.as_ref(),
                in_path,
                value_expr,
                borrowed_params,
                env,
                out,
            )?;
        }
        ast::TypeName::Writeable => todo!(),
        ast::TypeName::StrReference(..) => {
            gen_rust_reference_to_js(typ, in_path, value_expr, borrowed_params, env, out)?;
        }
        ast::TypeName::PrimitiveSlice(..) => todo!(),
        ast::TypeName::Unit => write!(out, "{}", value_expr)?,
    }

    Ok(())
}

/// TODO: docs
pub fn gen_rust_reference_to_js<W: fmt::Write>(
    underlying: &ast::TypeName,
    in_path: &ast::Path,
    value_expr: &str,
    borrowed_params: &ast::BorrowedParams,
    env: &Env,
    out: &mut W,
) -> fmt::Result {
    match underlying {
        ast::TypeName::Box(typ) | ast::TypeName::Reference(.., typ) => {
            gen_rust_reference_to_js(
                typ.as_ref(),
                in_path,
                &format!(
                    "(new Uint32Array(wasm.memory.buffer, {}, 1))[0]",
                    value_expr
                ),
                borrowed_params,
                env,
                out,
            )?;
        }
        ast::TypeName::Option(underlying) => match underlying.as_ref() {
            ast::TypeName::Box(..) | ast::TypeName::Reference(..) => write!(
                out,
                "{}",
                display::iife(|mut f| {
                    let ident = ast::Ident::from("option_ptr");
                    writeln!(
                        f,
                        "const {ident} = (new Uint32Array(wasm.memory.buffer, {}, 1))[0];",
                        value_expr
                    )?;
                    writeln!(
                        f,
                        "return ({ident} == 0) ? null : {};",
                        display::expr(|mut f| {
                            gen_value_rust_to_js(
                                ident.as_str(),
                                underlying.as_ref(),
                                in_path,
                                borrowed_params,
                                env,
                                &mut f,
                            )
                        })
                    )
                })
            )?,
            other @ (ast::TypeName::StrReference(..) | ast::TypeName::PrimitiveSlice(..)) => {
                panic!("`{}` is a fat pointer (ptr, length), and so it can't be stored behind an option.", other)
            }
            other => panic!(
                "`{0}` doesn't have the same alignment as ptr, so `Option<{0}>` is unsupported",
                other
            ),
        },

        ast::TypeName::Result(_, _) => {
            todo!("Receiving references to results")
        }

        ast::TypeName::Named(path_type) => {
            let custom_type = path_type.resolve(in_path, env);

            if let ast::CustomType::Enum(enm) = custom_type {
                write!(
                    out,
                    "{}_rust_to_js[{}]",
                    enm.name,
                    display::expr(|mut f| {
                        gen_rust_reference_to_js(
                            &ast::TypeName::Primitive(ast::PrimitiveType::isize),
                            in_path,
                            value_expr,
                            borrowed_params,
                            env,
                            &mut f,
                        )
                    })
                )?;
            } else {
                let params_borrowing_non_strrefs = borrowed_params
                    .borrowed_params
                    .iter()
                    .filter(|param| {
                        !param.ty.all_lifetimes(|lifetime, origin| {
                            matches!(
                                (lifetime, origin),
                                (ast::Lifetime::Named(_), ast::LifetimeOrigin::StrReference)
                            )
                        })
                    })
                    .collect::<Vec<_>>();

                if !borrowed_params.borrows_self && params_borrowing_non_strrefs.is_empty() {
                    write!(out, "new {}({})", custom_type.name(), value_expr)?;
                } else {
                    write!(
                        out,
                        "{}",
                        display::iife(|mut f| {
                            writeln!(f, "const out = new {}({});", custom_type.name(), value_expr)?;
                            if borrowed_params.borrows_self {
                                writeln!(f, "out.__this_lifetime_guard = this;")?;
                            }
                            for param in params_borrowing_non_strrefs.iter() {
                                writeln!(f, "out.__{0}_lifetime_guard = {0};", param.name)?;
                            }
                            writeln!(f, "return out;")
                        })
                    )?;
                }
            }
        }

        ast::TypeName::Primitive(prim) => {
            let prim_type = match prim {
                ast::PrimitiveType::i8 => "Int8Array",
                ast::PrimitiveType::u8 => "Uint8Array",
                ast::PrimitiveType::i16 => "Int16Array",
                ast::PrimitiveType::u16 => "Uint16Array",
                ast::PrimitiveType::i32 => "Int32Array",
                ast::PrimitiveType::u32 => "Uint32Array",
                ast::PrimitiveType::i64 => "BigInt64Array",
                ast::PrimitiveType::u64 => "BigUint64Array",
                ast::PrimitiveType::i128 => panic!("i128 not supported on JS"),
                ast::PrimitiveType::u128 => panic!("u128 not supported on JS"),
                ast::PrimitiveType::isize => "Int32Array",
                ast::PrimitiveType::usize => "Uint32Array",
                ast::PrimitiveType::f32 => "Float32Array",
                ast::PrimitiveType::f64 => "Float64Array",
                ast::PrimitiveType::bool => {
                    return write!(
                        out,
                        "(new Uint8Array(wasm.memory.buffer, {}, 1))[0] == 1",
                        value_expr
                    )
                }
                ast::PrimitiveType::char => {
                    return write!(
                        out,
                        "String.fromCharCode((new Uint32Array(wasm.memory.buffer, {}, 1))[0])",
                        value_expr
                    )
                }
            };

            write!(
                out,
                "(new {}(wasm.memory.buffer, {}, 1))[0]",
                prim_type, value_expr
            )?;
        }

        ast::TypeName::Unit => {
            write!(out, "{{}}")?;
        }

        ast::TypeName::StrReference(..) => {
            write!(
                out,
                "{}",
                display::iife(|mut f| {
                    writeln!(
                        f,
                        "const [ptr, len] = new Uint32Array(wasm.memory.buffer, {}, 2);",
                        value_expr
                    )?;
                    writeln!(f, "return diplomatRuntime.readString(wasm, ptr, len);")
                })
            )?;
        }
        _ => todo!(),
    }

    Ok(())
}
