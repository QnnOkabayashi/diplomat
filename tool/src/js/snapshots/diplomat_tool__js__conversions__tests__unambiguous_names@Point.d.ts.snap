---
source: tool/src/js/conversions.rs
expression: out_texts.get(out).unwrap()
---

export class Point {
  x: number;
  y: number;
}

